File: src/navigate.zig
//! Navigation / query over an encoded struple buffer.
//!
//! A buffer is a stream of elements; these helpers slice and inspect it without
//! decoding values. Everything is zero-copy — results are sub-slices of the
//! input, each itself a valid struple buffer, so it all composes and recurses.
//! Mirrors the reader surface of the original `Tuple.h` (head/tail/at/count/
//! contained-items + `is*` predicates + map helpers).
//!
//! Stream ops (`count`, `at`, `head`, `tail`, `nthRest`, `take`) operate on the
//! buffer's top-level element stream. To descend into an array/map/set, use
//! `containedItems` to get its inner stream, then view that.

const std = @import("std");
const struple = @import("struple.zig");

const Reader = struple.Reader;
const tc = struple.tc;
const DecodeError = struple.DecodeError;
const AllocError = std.mem.Allocator.Error;

pub fn view(bytes: []const u8) View {
    return View.init(bytes);
}

pub const View = struct {
    bytes: []const u8,

    pub fn init(bytes: []const u8) View {
        return .{ .bytes = bytes };
    }

    pub fn reader(self: View) Reader {
        return Reader.init(self.bytes);
    }

    /// Number of top-level elements.
    pub fn count(self: View) DecodeError!usize {
        var r = self.reader();
        var n: usize = 0;
        while (try r.skip()) n += 1;
        return n;
    }

    /// The element at `index`, as a zero-copy sub-view (null if out of range).
    pub fn at(self: View, index: usize) DecodeError!?[]const u8 {
        var r = self.reader();
        var i: usize = 0;
        while (try r.nextView()) |v| : (i += 1) {
            if (i == index) return v;
        }
        return null;
    }

    /// The first element (null if empty).
    pub fn head(self: View) DecodeError!?[]const u8 {
        return self.at(0);
    }

    /// Everything after the first element (empty if 0 or 1 elements).
    pub fn tail(self: View) DecodeError![]const u8 {
        var r = self.reader();
        _ = try r.nextView();
        return r.rest();
    }

    /// Drop `n` elements; return the remaining stream.
    pub fn nthRest(self: View, n: usize) DecodeError![]const u8 {
        var r = self.reader();
        var i: usize = 0;
        while (i < n) : (i += 1) {
            if (!(try r.skip())) break;
        }
        return r.rest();
    }

    /// The first `n` elements, as a contiguous sub-view.
    pub fn take(self: View, n: usize) DecodeError![]const u8 {
        var r = self.reader();
        var i: usize = 0;
        while (i < n) : (i += 1) {
            if (!(try r.skip())) break;
        }
        return self.bytes[0 .. self.bytes.len - r.rest().len];
    }

    /// The type code of the first element (null if empty).
    pub fn headType(self: View) ?u8 {
        return if (self.bytes.len > 0) self.bytes[0] else null;
    }

    pub fn isNil(self: View) bool {
        return self.headType() == tc.nil;
    }
    pub fn isUndefined(self: View) bool {
        return self.headType() == tc.undef;
    }
    pub fn isBool(self: View) bool {
        const t = self.headType() orelse return false;
        return t == tc.bool_false or t == tc.bool_true;
    }
    pub fn isInt(self: View) bool {
        const t = self.headType() orelse return false;
        return t == tc.int_zero or t == tc.int_neg_big or t == tc.int_pos_big or
            (t >= tc.int_neg_min and t <= tc.int_neg_max) or
            (t >= tc.int_pos_min and t <= tc.int_pos_max);
    }
    pub fn isFloat(self: View) bool {
        const t = self.headType() orelse return false;
        return t == tc.float32 or t == tc.float64;
    }
    pub fn isDecimal(self: View) bool {
        return self.headType() == tc.decimal;
    }
    pub fn isNumber(self: View) bool {
        return self.isInt() or self.isFloat() or self.isDecimal();
    }
    pub fn isTimestamp(self: View) bool {
        return self.headType() == tc.timestamp;
    }
    pub fn isUuid(self: View) bool {
        return self.headType() == tc.uuid;
    }
    pub fn isString(self: View) bool {
        return self.headType() == tc.string;
    }
    pub fn isBytes(self: View) bool {
        return self.headType() == tc.bytes;
    }
    pub fn isArray(self: View) bool {
        return self.headType() == tc.array;
    }
    pub fn isMap(self: View) bool {
        return self.headType() == tc.map;
    }
    pub fn isSet(self: View) bool {
        return self.headType() == tc.set;
    }
    pub fn isContainer(self: View) bool {
        const t = self.headType() orelse return false;
        return t == tc.array or t == tc.map or t == tc.set;
    }

    /// The first element's framed body when it is a container (escapes intact,
    /// zero-copy). Null if the head isn't a container.
    pub fn containerBody(self: View) DecodeError!?[]const u8 {
        if (!self.isContainer()) return null;
        var r = self.reader();
        const e = (try r.next()) orelse return null;
        return switch (e) {
            .array, .map, .set => |b| b,
            else => null,
        };
    }

    /// The container's inner element stream, un-escaped (owned by the caller).
    /// View it with `struple.view`, or a map with `struple.MapView`.
    pub fn containedItems(self: View, allocator: std.mem.Allocator) (DecodeError || AllocError)!?[]u8 {
        const body = (try self.containerBody()) orelse return null;
        return try struple.unescapeAlloc(allocator, body);
    }
};

/// Reads key/value pairs from a map's *inner* stream (the un-escaped body from
/// `View.containedItems`). Keys are in canonical (sorted) order, so `get`
/// early-exits.
pub const MapView = struct {
    inner: []const u8,

    pub fn init(inner: []const u8) MapView {
        return .{ .inner = inner };
    }

    pub fn count(self: MapView) DecodeError!usize {
        return (try View.init(self.inner).count()) / 2;
    }

    pub const Entry = struct { key: []const u8, value: []const u8 };

    pub const Iterator = struct {
        r: Reader,
        pub fn next(self: *Iterator) DecodeError!?Entry {
            const k = (try self.r.nextView()) orelse return null;
            const v = (try self.r.nextView()) orelse return error.Truncated;
            return Entry{ .key = k, .value = v };
        }
    };

    pub fn iterator(self: MapView) Iterator {
        return .{ .r = Reader.init(self.inner) };
    }

    /// Look up `key` (the encoded bytes of a key element, e.g. from packing a
    /// value). Returns the value's encoded bytes, or null. Ordered scan with
    /// early exit thanks to canonical key order.
    pub fn get(self: MapView, key: []const u8) DecodeError!?[]const u8 {
        var it = self.iterator();
        while (try it.next()) |e| {
            switch (std.mem.order(u8, e.key, key)) {
                .eq => return e.value,
                .gt => return null,
                .lt => {},
            }
        }
        return null;
    }

    /// Materialize a random-access index for O(log n) `get` and O(1) `at` (see
    /// `IndexedMap`). One O(n) pass; the caller owns the result.
    pub fn indexed(self: MapView, allocator: std.mem.Allocator) (DecodeError || AllocError)!IndexedMap {
        return IndexedMap.init(allocator, self.inner);
    }
};

/// A map's entries materialized into a random-access index. Building it is one
/// O(n) pass over the inner stream; thereafter `get` is an O(log n) binary search
/// (canonical key order means a key memcmp *is* the sort order) and `at` is O(1).
///
/// Use `MapView` directly for a single lookup (zero-alloc); reach for `IndexedMap`
/// when you do many lookups, or need positional access, on the same map. The entry
/// slices borrow the inner stream, so keep it alive for the index's lifetime.
pub const IndexedMap = struct {
    entries: []MapView.Entry,

    /// Build the index from a map's *inner* stream (the un-escaped body from
    /// `View.containedItems`). Free it with `deinit`; keep `inner` alive meanwhile.
    pub fn init(allocator: std.mem.Allocator, inner: []const u8) (DecodeError || AllocError)!IndexedMap {
        var list = std.ArrayList(MapView.Entry).init(allocator);
        errdefer list.deinit();
        var r = Reader.init(inner);
        while (try r.nextView()) |k| {
            const v = (try r.nextView()) orelse return error.Truncated;
            try list.append(.{ .key = k, .value = v });
        }
        return .{ .entries = try list.toOwnedSlice() };
    }

    pub fn deinit(self: *IndexedMap, allocator: std.mem.Allocator) void {
        allocator.free(self.entries);
        self.entries = &.{};
    }

    /// Number of entries — O(1).
    pub fn count(self: IndexedMap) usize {
        return self.entries.len;
    }

    /// The entry at `index` in canonical (sorted) order — O(1); null if out of range.
    pub fn at(self: IndexedMap, index: usize) ?MapView.Entry {
        return if (index < self.entries.len) self.entries[index] else null;
    }

    /// Look up `key` (an encoded key element) — O(log n) binary search. Returns the
    /// value's encoded bytes, or null.
    pub fn get(self: IndexedMap, key: []const u8) ?[]const u8 {
        return if (self.find(key)) |i| self.entries[i].value else null;
    }

    /// The index of `key` in canonical order, or null — O(log n).
    pub fn find(self: IndexedMap, key: []const u8) ?usize {
        var lo: usize = 0;
        var hi: usize = self.entries.len;
        while (lo < hi) {
            const mid = lo + (hi - lo) / 2;
            switch (std.mem.order(u8, self.entries[mid].key, key)) {
                .eq => return mid,
                .lt => lo = mid + 1,
                .gt => hi = mid,
            }
        }
        return null;
    }

    pub const Iterator = struct {
        entries: []const MapView.Entry,
        i: usize = 0,
        pub fn next(self: *Iterator) ?MapView.Entry {
            if (self.i >= self.entries.len) return null;
            defer self.i += 1;
            return self.entries[self.i];
        }
    };

    /// Entries in canonical (sorted) order.
    pub fn iterator(self: IndexedMap) Iterator {
        return .{ .entries = self.entries };
    }
};

File: dart/lib/src/navigate.dart
// Navigation / query over an encoded struple buffer.
//
// A buffer is a stream of elements; these helpers slice and inspect it without
// decoding values. Everything is zero-copy — results are sub-slices of the
// input, each itself a valid struple buffer, so it all composes and recurses.
//
// Stream ops (count, at, head, tail, nthRest, take) operate on the buffer's
// top-level element stream. To descend into an array/map/set, use
// containedItems to get its inner stream, then view that.

import 'dart:typed_data';

import 'codec.dart';

/// A zero-copy inspector over an encoded struple buffer.
class View {
  final Uint8List bytes;

  View(this.bytes);

  Reader _reader() => Reader(bytes);

  /// The number of top-level elements.
  int count() {
    final r = _reader();
    var n = 0;
    while (r.skip()) {
      n++;
    }
    return n;
  }

  /// The element at [index] as a zero-copy sub-view, or null if out of range.
  Uint8List? at(int index) {
    final r = _reader();
    var i = 0;
    Uint8List? v;
    while ((v = r.nextView()) != null) {
      if (i == index) return v;
      i++;
    }
    return null;
  }

  /// The first element, or null if empty.
  Uint8List? head() => at(0);

  /// Everything after the first element (empty if 0 or 1 elements).
  Uint8List tail() {
    final r = _reader();
    r.nextView();
    return r.rest();
  }

  /// Drops [n] elements and returns the remaining stream.
  Uint8List nthRest(int n) {
    final r = _reader();
    for (var i = 0; i < n; i++) {
      if (!r.skip()) break;
    }
    return r.rest();
  }

  /// The first [n] elements as a contiguous sub-view.
  Uint8List take(int n) {
    final r = _reader();
    for (var i = 0; i < n; i++) {
      if (!r.skip()) break;
    }
    return Uint8List.sublistView(bytes, 0, bytes.length - r.rest().length);
  }

  /// The type code of the first element, or null if empty.
  int? headType() => bytes.isNotEmpty ? bytes[0] : null;

  bool isNil() => headType() == TypeCode.nil;
  bool isUndefined() => headType() == TypeCode.undef;

  bool isBool() {
    final t = headType();
    return t == TypeCode.boolFalse || t == TypeCode.boolTrue;
  }

  bool isInt() {
    final t = headType();
    if (t == null) return false;
    return t == TypeCode.intZero ||
        t == TypeCode.intNegBig ||
        t == TypeCode.intPosBig ||
        (t >= TypeCode.intNegMin && t <= TypeCode.intNegMax) ||
        (t >= TypeCode.intPosMin && t <= TypeCode.intPosMax);
  }

  bool isFloat() {
    final t = headType();
    return t == TypeCode.float32 || t == TypeCode.float64;
  }

  bool isDecimal() => headType() == TypeCode.decimal;

  bool isNumber() => isInt() || isFloat() || isDecimal();

  bool isTimestamp() => headType() == TypeCode.timestamp;
  bool isUuid() => headType() == TypeCode.uuid;
  bool isString() => headType() == TypeCode.string;
  bool isBytes() => headType() == TypeCode.bytes;
  bool isArray() => headType() == TypeCode.array;
  bool isMap() => headType() == TypeCode.map;
  bool isSet() => headType() == TypeCode.set;

  bool isContainer() {
    final t = headType();
    return t == TypeCode.array || t == TypeCode.map || t == TypeCode.set;
  }

  /// The first element's framed body when it is a container (escapes intact,
  /// zero-copy), or null if the head isn't a container.
  Uint8List? containerBody() {
    if (!isContainer()) return null;
    final e = _reader().next();
    if (e == null) return null;
    switch (e.kind) {
      case Kind.array:
      case Kind.map:
      case Kind.set:
        return e.body;
      default:
        return null;
    }
  }

  /// The container's inner element stream, un-escaped (a fresh slice). View it
  /// with [View], or a map with [MapView]. Null if the head isn't a container.
  Uint8List? containedItems() {
    final body = containerBody();
    if (body == null) return null;
    return unescape(body);
  }
}

// ---------------------------------------------------------------------------
// MapView — reads key/value pairs from a map's inner stream
// ---------------------------------------------------------------------------

/// A key/value pair of encoded element views.
class MapEntry {
  final Uint8List key;
  final Uint8List value;
  const MapEntry(this.key, this.value);
}

/// Reads key/value pairs from a map's inner stream (the un-escaped body from
/// [View.containedItems]). Keys are in canonical (sorted) order, so [get]
/// early-exits.
class MapView {
  final Uint8List inner;
  MapView(this.inner);

  /// The number of entries.
  int count() => View(inner).count() ~/ 2;

  /// An iterator over the map's entries in canonical order.
  MapIterator iterator() => MapIterator(Reader(inner));

  /// Looks up [key] (the encoded bytes of a key element). Returns the value's
  /// encoded bytes, or null. Ordered scan with early exit.
  Uint8List? get(Uint8List key) {
    final it = iterator();
    MapEntry? e;
    while ((e = it.next()) != null) {
      final c = compareBytes(e!.key, key);
      if (c == 0) return e.value;
      if (c > 0) return null;
    }
    return null;
  }

  /// Materializes a random-access index for O(log n) get and O(1) at.
  IndexedMap indexed() => IndexedMap(inner);
}

/// Yields entries of a [MapView] in canonical order.
class MapIterator {
  final Reader _r;
  MapIterator(this._r);

  /// The next entry, or null at end of stream.
  MapEntry? next() {
    final k = _r.nextView();
    if (k == null) return null;
    final v = _r.nextView();
    if (v == null) throw const StrupleException('truncated input');
    return MapEntry(k, v);
  }
}

// ---------------------------------------------------------------------------
// IndexedMap — random-access index over a map's entries
// ---------------------------------------------------------------------------

/// Materializes a map's entries into a random-access index. Building it is one
/// O(n) pass over the inner stream; thereafter [get] is an O(log n) binary
/// search (canonical key order means a key compare is the sort order) and [at]
/// is O(1). The entry slices borrow the inner stream, so keep it alive.
class IndexedMap {
  final List<MapEntry> entries;

  IndexedMap._(this.entries);

  /// Builds an index from a map's inner stream (the un-escaped body from
  /// [View.containedItems]).
  factory IndexedMap(Uint8List inner) {
    final list = <MapEntry>[];
    final r = Reader(inner);
    Uint8List? k;
    while ((k = r.nextView()) != null) {
      final v = r.nextView();
      if (v == null) throw const StrupleException('truncated input');
      list.add(MapEntry(k!, v));
    }
    return IndexedMap._(list);
  }

  /// The number of entries — O(1).
  int count() => entries.length;

  /// The entry at [index] in canonical order — O(1), or null if out of range.
  MapEntry? at(int index) {
    if (index < 0 || index >= entries.length) return null;
    return entries[index];
  }

  /// Looks up [key] (an encoded key element) — O(log n). Returns the value's
  /// encoded bytes, or null.
  Uint8List? get(Uint8List key) {
    final i = find(key);
    return i == null ? null : entries[i].value;
  }

  /// The index of [key] in canonical order — O(log n), or null if absent.
  int? find(Uint8List key) {
    var lo = 0;
    var hi = entries.length;
    while (lo < hi) {
      final mid = lo + (hi - lo) ~/ 2;
      final c = compareBytes(entries[mid].key, key);
      if (c == 0) return mid;
      if (c < 0) {
        lo = mid + 1;
      } else {
        hi = mid;
      }
    }
    return null;
  }

  /// The entries in canonical (sorted) order.
  Iterable<MapEntry> get iterable => entries;
}
